rllib/core/learner/learner_group.py

```diff
@@ -175,60 +188,98 @@ def update(
-                train_batch[module_id] = batch.policy_batches[module_id]
-        train_batch = MultiAgentBatch(train_batch, batch.count)
+        # TODO (sven): Move this into individual Learners. It might be that
+        #  batch/episodes postprocessing on each Learner requires the non-trainable
+        #  modules' data.
+        train_batch = None
+        if batch is not None:
+            train_batch = {}
+            for module_id in batch.policy_batches.keys():
+                if self._is_module_trainable(module_id, batch):
+                    train_batch[module_id] = batch.policy_batches[module_id]
+            train_batch = MultiAgentBatch(train_batch, batch.count)
 
         if self.is_local:
+            assert batch is not None or episodes is not None
             results = [
                 self._learner.update(
-                    train_batch,
+                    batch=train_batch,
+                    episodes=episodes,
                     minibatch_size=minibatch_size,
                     num_iters=num_iters,
                     reduce_fn=reduce_fn,
                 )
             ]
         else:
 
-            def _learner_update(learner, minibatch):
+            def _learner_update(learner: Learner, batch_shard=None, episodes_shard=None):
                 return learner.update(
-                    minibatch,
+                    batch=batch_shard,
+                    episodes=episodes_shard,
                     minibatch_size=minibatch_size,
                     num_iters=num_iters,
                     reduce_fn=reduce_fn,
                 )
 
-            results = self._get_results(
-                self._worker_manager.foreach_actor(
-                    [
-                        partial(_learner_update, minibatch=minibatch)
-                        for minibatch in ShardBatchIterator(batch, len(self._workers))
-                    ]
+            # Only batch provided, split it up into n shards.
+            if episodes is None:
+                assert batch is not None
+                results = self._get_results(
+                    self._worker_manager.foreach_actor(
+                        [
+                            partial(_learner_update, batch_shard=batch_shard)
+                            for batch_shard in ShardBatchIterator(train_batch, len(self._workers))
+                        ]
+                    )
                 )
-            )
+            elif batch is None:
+                assert episodes is not None
+                results = self._get_results(
+                    self._worker_manager.foreach_actor(
+                        [
+                            partial(_learner_update, episodes_shard=episodes_shard)
+                            for episodes_shard in ShardEpisodesIterator(episodes, len(self._workers))
+                        ]
+                    )
+                )
+            # TODO (sven): Implement the case in which both batch and episodes might
```

Comment:
If we have mutually exclusive methods: `update_from_batch` and `update_from_episodes`, this case (both batch AND episodes provided by user) would not exist anyways.